03/ukol16.py

```diff
@@ -0,0 +1,22 @@
+#Napiš program, který postupně načte od uživatele dvě čísla a jednoznakový řetězec – buď '+', '-', '*'
+#nebo '/'. Program provede na číslech příslušnou operaci.
+
+prvni = int(input('Zadej prvni cislo: '))
```

Comment:
```
Zadej prvni cislo: Ahoj
Traceback (most recent call last):
  File "03/ukol16.py", line 4, in <module>
    prvni = int(input('Zadej prvni cislo: '))
ValueError: invalid literal for int() with base 10: 'Ahoj'
```

😢 

Jakože jo, člověk by měl zadat číslo _(a aj je mu to výslovně řečeno)_ - ale pistoli mu u hlavy nikdo nedrží a vlastně si tam může napsat, co chce - takže ideálně by bylo fajn mít vstup (resp. vstupy) ošetřené.

Třeba tak _(pokud na to nechceme používat [výjimky](https://docs.python.org/3/tutorial/errors.html) - respektive jejich odchytávání)_, že budeš kontroloval, zda ten string prostě fakt obsahuje číslo.

Typ `str` (string) samotný na to má vhodné metody:
- `str.replace(str_x, str_y, int_z)` (https://docs.python.org/3/library/stdtypes.html#str.replace) - vrátí nový string, který vznikne nahrazením výskytů `str_x` za `str_y`, ale maximálně `int_z`-krát. Tohle můžeme použít k ošetření vstupu, kde je desetinná tečka _(proč bychom uživateli zakazovali sčítat desetinná čísla, žejo)_
  - Příklady:
    - `"ahojky mňauky vole".replace("ky", "LOL")` -> `"ahojLOL mňauLOL vole"` (třetí argument - limit - jsme nezadali, takže se nahradí šecko.
    - `"ahojky mňauky vole".replace("ky", "LOL", 1)` -> `"ahojLOL mňauky vole"` (třetí argument - limit - jsme zadali jako 1, takže se nahradí jen první `ky`.
  - ***My uděláme to, že nahradíme tečku `.` za nic ``, a to maximálně jednou: `str.replace(',', '', 1)`*** _(když by tam bylo víc teček, tak to stejně není validní číslo, žejo...)_
- `str.isdigit()` (https://docs.python.org/3/library/stdtypes.html#str.isdigit) - vrátí `True`, pokud string obsahuje pouze čísla.
  - Příklady:
    - `""` -> `False`
    - `"1"` -> `True`
    - `"8643500"` -> `True`
    - `"1.123"` -> `False` _(páč je tam tečka!)_

A tím, že tyhle dvě metody zkombinujeme, tak zařídíme to, co potřebujeme - zkontrolujeme, zda string obsahuje celé číslo a nebo desetinné číslo _(stačí nám vědět, že je to jedno z toho)_.

```
`"".replace('.', '', 1).isdigit()` -> `False`
`"123".replace('.', '', 1).isdigit()` -> `True`
`"123.456".replace('.', '', 1).isdigit()` -> `True`  # Páč se nejdřív dá pryč první tečka a pak kontrolujeme, zda výsledek ("123456") obsahuje jen čísla.
`"123.456.789".replace('.', '', 1).isdigit()` -> `False`  # Páč po odstranění první tečky zbyte "123456.789" a to NEobsahuje jen čísla.
```

A pak user input čísel vyřešíme takto:

```
# Zkontrolujeme vstup validního čísla - první číslo.
while True:
  prvni = input('Dej první: ')
  if prvni.replace('.', '',1).isdigit():
    break

# Zkontrolujeme vstup validního znaménka.
while True:
  znamenko = input('Dej znaménko: ')
  if znamenko in "+-/*":  
    break

# Zkontrolujeme vstup validního čísla - první číslo.
while True:
  druhe = input('Dej první: ')
  if druhe.replace('.', '',1).isdigit():
    break
```

Reply:
Btw. důvod, proč na konci ještě řešíme převod z `float` na `int` je vlastně nakonec čistě estetický. Kdyby totiž aj celá čísla byla v datovém typu `float` a ne `int`, tak je Python pak v printu vypisuje s desetinnou čárkou (tečkou), nějak takto:

```
Dej první: 5
Dej znaménko: *
Dej první: 3
5.0 * 3.0 = 15.0
```

_(všimni si těch vlastně zbytečných `.0`...)_